src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.Browser.cs

```diff
@@ -0,0 +1,31 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.IO;
+using System.Diagnostics;
+
+namespace System.Security.Cryptography
+{
+    internal partial class RandomNumberGeneratorImplementation
+    {
+        private static readonly FileStream s_randomStream = new FileStream("/dev/random", FileMode.Open, FileAccess.Read);
+
+        private static unsafe void GetBytes(byte* pbBuffer, int count)
+        {
+            Debug.Assert(count > 0);
+
+            int pos = 0;
+            while (pos < count)
+            {
+                var span = new Span<byte>(pbBuffer + pos, count - pos);
+                int res = s_randomStream.Read(span);
```

Comment:
Such non-thread-safe usage is ok because GetBytes will only ever be called from a single thread at a time by the nature of wasm?  Presumably at some point multi-threading will end up being supported; we'll deal with that here and everywhere else when that happens?  Worth tracking in some way?

Reply:
When running under emscripten, reads from /dev/random are redirected to JS code which calls Crypto.getRandomValues () to get the random data. So that part should be thread safe.

Reply:
If this could be a problem, I can add a SystemNative function instead to skip the managed code.

Reply:
FileStream itself isn't.  For example, it maintains its own buffer, and if the passed in span is smaller than its own buffer, it's going to read into its buffer (updating fields like how much data is available in the buffer and where it starts) and then populate the span from that buffer.  Multiple threads accessing this concurrently could hit various issues, like the same data being returned to two different threads, those buffer offsets/lengths being updated in an inconsistent manner, etc.

Reply:
At a minimum I'd suggest initializing the FileStream with a buffer size of 1, which means it'll skip its buffer entirely.  That's probably good for other reasons as well.  You could then lock (s_randomStream) around this to ensure thread-safe access.  I don't have a sense for how hot-path this might be to know whether it's actually an issue, but assuming single-threaded access in general, the overhead of such a lock here should be minimal.

Reply:
> I can add a SystemNative function instead to skip the managed code.

Maybe just call `SystemNative_GetNonCryptographicallySecureRandomBytes` with a comment that it returns  cryptographic strong random bytes on Wasm?

Reply:
GetNonCryptographicallySecureRandomBytes  can silently fall back to a PRNG if /dev/urandom cannot be opened etc.

Reply:
Ok, I forgot about that. We can a second function like this without the fallbacks or consider deleting the fallbacks - they should not be needed on any platform we care about.